File: nrs/core/base/Link.java
/*
 * Copyright (C) 2004 Edinburgh University
 *
 *    This program is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation; either version 2 of
 *    the License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public
 *    License along with this program; if not, write to the Free
 *    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *    MA 02111-1307 USA
 *
 * For further information in the first instance contact:
 * Richard Reeve <richardr@inf.ed.ac.uk>
 *
 */
package nrs.core.base;

import nrs.core.base.Message;
import nrs.core.base.Variable;
import nrs.core.message.DeleteLink;
import nrs.core.message.Constants;

/** Class for representing Links between NRS variables.
 *
 * @author Thomas French
 */

public class Link {

    private String m_sourceCID;
    private Integer m_sourceVNID;

    private String m_targetCID;
    private Integer m_targetVNID;
    private String m_targetVNName;

    //private String m_logPort;
    private boolean m_temporary;
    private String m_route;

    private Variable m_sourceRef = null;
    private Variable m_targetRef = null;

    private RouteManager m_routeMan;

    private boolean m_isSource = false;

    /**
     * Constructor for capturing all the CreateLink message information.
     * Using VNIDs for source and target.
     *
     * @param sourceCID
     * @param sourceVNID
     * @param targetCID
     * @param targetVNID
     * @param temporary
     * @param route
     * @param rMan 
     * @param myCID
     *
    */
    public Link(String sourceCID, Integer sourceVNID, String targetCID,
		Integer targetVNID, boolean temporary, String route, 
		RouteManager rMan, String myCID){

	m_sourceCID = sourceCID;
	m_sourceVNID = sourceVNID;
	m_targetCID = targetCID;
	m_targetVNID = targetVNID;
	//m_logPort = logPort;
	m_temporary = temporary;
	m_route = route;
	m_routeMan = rMan;

	if ( sourceCID.equals(myCID) )
	    m_isSource = true; 
    }

    /**
     * Constructor for capturing all the CreateLink message information.
     * Using VNID for source and VNName for target.
     *
     * @param sourceCID
     * @param sourceVNID
     * @param targetCID
     * @param targetVNName
     * @param temporary
     * @param route
     * @param rMan 
     *
     * @deprecated outdated, do NOT use.
    */
    public Link(String sourceCID, Integer sourceVNID, String targetCID,
		String targetVNName, boolean temporary, String route, 
		RouteManager rMan){
	m_sourceCID = sourceCID;
	m_sourceVNID = sourceVNID;
	m_targetCID = targetCID;
	m_targetVNName = targetVNName;
	//m_logPort = logPort;
	m_temporary = temporary;
	m_route = route;
	m_routeMan = rMan;

	//must be 0 if not supplied
	m_targetVNID = new Integer(0);
    }

    /** Constructor, case where link is onboard component. */
    public Link(Variable s, Variable t){
	m_sourceRef = s;
	m_targetRef = t;
    }

    //Accessor methods
    // getters

    public String getSourceCID(){
	return m_sourceCID;
    }
    public String getTargetCID(){
	return m_targetCID;
    }

    public Integer getSourceVNID(){
	if ( m_sourceVNID != null ) return m_sourceVNID;
	return null;
    }

    public Integer getTargetVNID(){
	if ( m_targetVNID != null ) return m_targetVNID;
	return null;
    }

    public boolean isSource(int vnid){
	return ( m_isSource || 
		 ( m_sourceRef != null && m_sourceRef.getVNID() == vnid ) );
    }

    /** 
     * @deprecated 
     */
    public String getTargetVNName(){
	if ( m_targetVNName != null ) return m_targetVNName;
	return null;
    }

    public boolean getTemporary(){
	return m_temporary;
    }

    /** For on-board links. 
     * @return Reference to source of link.
     */
    public Variable getSourceRef(){
	return m_sourceRef;
    }

    /** For on-board links. 
     * @return Reference to target of link.
     */
    public Variable getTargetRef(){
	return m_targetRef;
    }

    //-----------------------------------------------------------------

    /** Fill information in to Message object from Link information.
     *
     * @param msg {@link Message} object to add information to.
    */
    public void fillMsg(Message msg) {

	// Add route
	m_route = m_routeMan.resolveRoute(m_targetCID, true);
	msg.setNRSField(Constants.MessageFields.route, m_route);
	
	// Add target
	if (m_targetVNID != null)
	    {
		msg.setNRSField(Constants.MessageFields.toVNID,
				m_targetVNID.toString());
	    }
	
	// Fix route and target
	MessageTools.fixRoute(msg, m_targetCID);
	MessageTools.fixToVNID(msg, m_targetVNName);
    }

    /** Fill in information for a DeleteLink message from Link information.
     *
     * @param msg Message to fill in.
     * @param source is caller source of link
     */
    public void fillDeleteLinkMsg(DeleteLink msg, boolean source){
	//fill in basic information required
	if ( source ) 
	    m_route = m_routeMan.resolveRoute(m_targetCID, true);
	else
	    m_route = m_routeMan.resolveRoute(m_sourceCID, true);

	msg.setNRSField(Constants.MessageFields.route, m_route);
	
	// Fix route of message and set sourceNotTarget flag
	if ( source ){
	    MessageTools.fixRoute(msg, m_targetCID);
	    msg.setField(Constants.Fields.F_sourceNotTarget, "false");
	}
	else{
	    MessageTools.fixRoute(msg, m_sourceCID);
	    msg.setField(Constants.Fields.F_sourceNotTarget, "true");
	}

	MessageTools.fixToVNID(msg, Constants.MessageTypes.DeleteLink);
	
	//fill in rest
	msg.setField(Constants.MessageFields.cid, m_sourceCID);
	msg.setField(Constants.MessageFields.vnid, m_sourceVNID.toString());	

	if ( m_targetVNID != null )
	    msg.setField(Constants.MessageFields.targetVNID, 
			 m_targetVNID.toString());

	msg.setField(Constants.MessageFields.targetCID, m_targetCID);
    }
}

File: Base/Link.cc
/*
 * Copyright (C) 2005 Richard Reeve and Edinburgh University
 *
 *    This program is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation; either version 2 of
 *    the License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public
 *    License along with this program; if not, write to the Free
 *    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *    MA 02111-1307 USA
 *
 * For further information in the first instance contact:
 * Richard Reeve <richardr@inf.ed.ac.uk>
 *
 */

#include "Encoding.hh"
#include "Exception.hh"
#include "IntelligentStore.hh"
#include "Link.hh"
#include "Message.hh"
#include "Segment.hh"
#include "Location.hh"

NRS::Base::Link::Link( const VariableManager &theVMRef, integer_t aVNID ) :
  iEncoding( theVMRef.getMessageDescription() ),
  iVM( theVMRef ), iInstantTransmit( false ),
  iReceivedMessage( false ),
  iMinIn( 0 ), iMaxIn( 0 ), iNumIn( 0 ),
  iMinOut( 0 ), iMaxOut( 0 ), iNumOut( 0 ),
  iMinLog( 0 ), iMaxLog( 0 ), iNumLog( 0 ),
  iVNID( aVNID )
{
}

NRS::Base::Link::~Link()
{
  _DEBUG_( "Destroying links" );
  if (iVNID > 0)
    {
      _DEBUG_( "Deleting links: " << iSource.size() );
      for ( unsigned i = 0; i < iSource.size(); ++i )
	{
          _DEBUG_( "iSource[i].iVNID = " << iSource[i].getVNID() ); 
	  Target aTarget( iSource[i], &iVM.getMessageDescription() );
	  aTarget.deleteLink( iVNID );
	}
      for ( unsigned i = 0; i < iMessagePtr.size(); ++i )
	iMessagePtr[i]->getTarget().deleteLink( iVNID );
    }
  _DEBUG_( "Deleting message pointers: " << iMessagePtr.size() );
  for ( unsigned i = 0; i < iMessagePtr.size(); ++i )
    delete iMessagePtr[i];
}

void NRS::Base::Link::receiveMessage( const MessageStore &aMessageStore,
				      const IntelligentStore *intelligentPtr )
{
  iEncoding.setMessage( aMessageStore, intelligentPtr );

  receiveMessage();
}

void NRS::Base::Link::receiveMessage()
{
  update();

  //if (iInstantTransmit)
  sendMessages();
}

void NRS::Base::Link::setInstantTransmission( bool instantTransmit )
{
  iInstantTransmit = instantTransmit;
}

void NRS::Base::Link::addSource( const Location &aSourceRef )
{
  iSource.push_back( aSourceRef );
  ++iNumIn;
}

bool NRS::Base::Link::removeSource( const Location &aSourceRef )
{
  _DEBUG_( "Removing some source" );
  typeof( iSource.begin() ) iter;

  for ( iter = iSource.begin();
	(iter != iSource.end()) && (aSourceRef != *iter);
	++iter );

  if (iter != iSource.end())
    {
      --iNumIn;
      iSource.erase( iter );
      return true;
    }
  return false;
}

void NRS::Base::Link::addTarget( const Target &aTarget )
{
  const VariableManager *aVMPtr = &iVM;
  bool found = false;
  const std::string &msgName = aTarget.getMessageDescriptionPtr()->iName;

  while (!found && (aVMPtr))
    {
      _DEBUG_( "testing " << aVMPtr->getMessageDescription().iName
	       << "==" << msgName );
      if (aVMPtr->getMessageDescription().iName != msgName)
	aVMPtr = aVMPtr->getParent();
      else
	found = true;
    }

  if (!found)
    _ABORT_( "Trying to connect to illegal destination" );
  
  ++iNumOut;

  iMessagePtr.push_back( new Message( aTarget ) );
  if (!aTarget.isLocal())
    iMessagePtr.back()->setEncodingStore( &iEncoding );
}

bool NRS::Base::Link::removeTarget( const Target &aTarget )
{
  _DEBUG_( "Removing some target" );
  typeof( iMessagePtr.begin() ) iter = iMessagePtr.begin();

  while ((iter != iMessagePtr.end()) && (aTarget != (*iter)->getTarget()))
    ++iter;
  
  if (iter != iMessagePtr.end())
    {
      --iNumOut;
      delete *iter;
      iMessagePtr.erase( iter );
      return true;
    }
  return false;
}

void NRS::Base::Link::addLog( Log *aLogPtr )
{
  ++iNumLog;
  _BUG_( "Unimplemented" );
}

bool NRS::Base::Link::removeLog( const Log *aLogPtr )
{
  --iNumLog;
  _BUG_( "Unimplemented" );
  return false;
}

bool NRS::Base::Link::resolveDependencies()
{
  if (iMinIn > iNumIn)
    return false;
  if ((iMaxIn != -1) && (iMaxIn < iNumIn))
    return false;
  if (iMinOut > iNumOut)
    return false;
  if ((iMaxOut != -1) && (iMaxOut < iNumOut))
    return false;
  if (iMinLog > iNumLog)
    return false;
  if ((iMaxLog != -1) && (iMaxLog < iNumLog))
    return false;

  return true;
}

void NRS::Base::Link::setConnections( int min_in, int max_in,
				      int min_out, int max_out,
				      int min_log, int max_log )
{
  iMinIn = min_in;
  iMaxIn = max_in;
  iMinOut = min_out;
  iMaxOut = max_out;
  iMinLog = min_log;
  iMaxLog = max_log;
}

void NRS::Base::Link::update()
{
  iReceivedMessage = true;
  iEncoding.update();
}

void NRS::Base::Link::sendMessages()
{
  if (iReceivedMessage)
    {
      _DEBUG_( "link start" );
      iReceivedMessage = false;
      alertObservers();
      for ( unsigned i = 0; i < iMessagePtr.size(); ++i )
	iMessagePtr[i]->sendMessages();
      _DEBUG_( "link end" );
    }
}
